LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDevice.cs

```diff
@@ -421,6 +431,78 @@ internal async Task<bool> UpdateAfterJoinAsync(string devAddr, string nwkSKey, s
             return succeeded;
         }
 
+        internal void SetRequestHandler(ILoRaDataRequestHandler dataRequestHandler) => this.dataRequestHandler = dataRequestHandler;
+
+        public void Queue(LoRaRequest request)
+        {
+            // Access to runningRequest and queuedRequests must be
+            // thread safe
+            lock (this)
+            {
+                if (this.runningRequest == null)
+                {
+                    this.StartNextRequest(request);
+                }
+                else
+                {
+                    this.queuedRequests.Enqueue(request);
+                }
+            }
+        }
+
+        private void QueueNext()
+        {
+            // Access to runningRequest and queuedRequests must be
+            // thread safe
+            lock (this)
+            {
+                this.runningRequest = null;
+                if (this.queuedRequests.TryDequeue(out var nextRequest))
+                {
+                    this.StartNextRequest(nextRequest);
+                }
+            }
+        }
+
+        void StartNextRequest(LoRaRequest requestToStart)
+        {
+            async Task RunAndQueueNext(LoRaRequest request)
+            {
+                LoRaDeviceRequestProcessResult result = null;
+                Exception processingError = null;
+
+                try
+                {
+                    result = await this.dataRequestHandler.ProcessRequestAsync(request, this);
```

Comment:
Here you are still holding onto the lock and pass control to an external method you don't control. If the ProcessRequestAsync would acquire another lock you can end up in a dead lock scenario. It's only a theoretical problem here.

Reply:
I prefer having the StartNextRequest doing the 2 lines (assign and schedule current task) than duplicating the code in the 2 places where it is called.

Reply:
Sure. Another Q: Why did you make RunAndQueueNext a local function? Just readability?

Reply:
Wanted to ensure that is it only called within the method